cluster/node.go

```diff
@@ -186,66 +188,59 @@ func (n *Node) refreshContainer(ID string) error {
 
 	if len(containers) > 1 {
 		// We expect one container, if we get more than one, trigger a full refresh.
-		return n.refreshContainers()
+		return n.RefreshContainers(full)
 	}
 
-	n.Lock()
-	defer n.Unlock()
-
 	if len(containers) == 0 {
 		// The container doesn't exist on the node, remove it.
+		n.Lock()
 		delete(n.containers, ID)
+		n.Unlock()
+
 		return nil
 	}
 
-	_, err = n.updateContainer(containers[0], n.containers)
+	_, err = n.updateContainer(containers[0], n.containers, full)
 	return err
 }
 
-func (n *Node) ForceRefreshContainer(c dockerclient.Container) error {
-	return n.inspectContainer(c, n.containers, true)
-}
+func (n *Node) updateContainer(c dockerclient.Container, containers map[string]*Container, full bool) (map[string]*Container, error) {
+	var container *Container
 
-func (n *Node) inspectContainer(c dockerclient.Container, containers map[string]*Container, lock bool) error {
-
-	container := &Container{
-		Container: c,
-		Node:      n,
+	if current, exists := n.containers[c.Id]; exists {
+		// The container is already known.
+		container = current
+	} else {
+		// This is a brand new container. We need to do a full refresh.
+		container = &Container{
+			Node: n,
+		}
+		full = true
 	}
 
-	info, err := n.client.InspectContainer(c.Id)
-	if err != nil {
-		return err
+	// Update its internal state.
+	container.Container = c
+
+	// Update ContainerInfo.
+	if full {
+		info, err := n.client.InspectContainer(c.Id)
+		if err != nil {
+			return nil, err
+		}
+		container.Info = *info
 	}
-	container.Info = *info
 
 	// real CpuShares -> nb of CPUs
 	container.Info.Config.CpuShares = container.Info.Config.CpuShares / 100.0 * n.Cpus
 
-	if lock {
-		n.Lock()
-		defer n.Unlock()
-	}
+	n.Lock()
+	defer n.Unlock()
 	containers[container.Id] = container
 
-	return nil
-}
-
-func (n *Node) updateContainer(c dockerclient.Container, containers map[string]*Container) (map[string]*Container, error) {
-	if current, exists := n.containers[c.Id]; exists {
-		// The container exists. Update its state.
-		current.Container = c
-		containers[current.Id] = current
-	} else {
-		// This is a brand new container.
-		if err := n.inspectContainer(c, containers, false); err != nil {
-			return nil, err
-		}
-	}
 	return containers, nil
 }
 
-func (n *Node) refreshContainersAsync() {
+func (n *Node) RefreshContainersAsync() {
```

Comment:
It's just to be consistent - since `RefreshContainer()` is now "public", I've made `RefreshContainers()` and `RefreshContainersAsync` public as well.